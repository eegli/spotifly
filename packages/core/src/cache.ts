/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable  @typescript-eslint/no-empty-function */

export enum CacheEntity {
  Artist = 'artist',
  Track = 'track',
}

export function Cacheable<T extends unknown[]>(
  entity: CacheEntity,
  debug = false
) {
  return function cacheMeBaby(
    _: unknown,
    propertyKey: string,
    descriptor: PropertyDescriptor
  ) {
    const log = debug ? console.log : () => {};
    const original = descriptor.value;
    descriptor.value = async function (...args: string[]) {
      if (!Array.isArray(args) || !args.every(arg => typeof arg === 'string')) {
        throw new Error(
          `Decorator applied on uncacheable method ${propertyKey}`
        );
      }
      const { cached, uncached } = cache.filterCached(entity, args);

      if (uncached.length === 0) {
        log('Reading all items from cache');
        return Promise.resolve(cached);
      }

      log(
        `Found ${cached.length}/${args.length} entries in cache, optimizing query...`
      );
      args = uncached;

      const result: T[] = await original.call(this, ...args);

      log(`Putting ${result.length} new item(s) into cache`);
      for (let i = 0; i < result.length; i++) {
        cache.put(entity, [uncached[i], result[i]]);
      }
      log(`Appending ${cached.length} item(s) to result from cache`);
      result.push(...cached);
      return Promise.resolve(result);
    };
  };
}

export type FilteredCache = {
  cached: any[];
  uncached: string[];
};

class Cache {
  #tracks = new Map<string, any>();
  #artists = new Map<string, any>();

  #enabled = true;

  #findBin(entity: CacheEntity) {
    switch (entity) {
      case CacheEntity.Track:
        return this.#tracks;
      case CacheEntity.Artist:
        return this.#artists;
    }
  }

  public disable() {
    this.#enabled = false;
  }

  public put(entity: CacheEntity, item: [string, any]): void {
    const bin = this.#findBin(entity);
    bin.set(item[0], item[1]);
  }

  public filterCached(entity: CacheEntity, ids: string[]): FilteredCache {
    if (!this.#enabled) return { cached: [], uncached: ids };
    const bin = this.#findBin(entity);
    return ids.reduce(
      (acc, curr) => {
        const item = bin.get(curr);
        if (item) {
          acc.cached.push(item);
        } else {
          acc.uncached.push(curr);
        }
        return acc;
      },
      {
        cached: [],
        uncached: [],
      } as FilteredCache
    );
  }
}

export const cache = new Cache();
