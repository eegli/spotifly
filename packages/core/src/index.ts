import { AuthProvider, AuthProviderOptions } from './provider';

import Albums from './models/albums';
import Artists from './models/artists';
import Categories from './models/categories';
import Episodes from './models/episodes';
import future from './models/future';
import Genres from './models/genres';
import Markets from './models/markets';
import Player from './models/player';
import Playlists from './models/playlists';
import Search from './models/search';
import Shows from './models/shows';
import Tracks from './models/tracks';
import Users from './models/users';

export function initialize(authOptions: AuthProviderOptions) {
  const provider = new AuthProvider(authOptions);
  return {
    Albums: Albums(provider),
    Artists: Artists(provider),
    Categories: Categories(provider),
    Episodes: Episodes(provider),
    Genres: Genres(provider),
    Markets: Markets(provider),
    Player: Player(provider),
    Playlists: Playlists(provider),
    Search: Search(provider),
    Shows: Shows(provider),
    Tracks: Tracks(provider),
    Users: Users(provider),
    future: future(provider),
  } as const;
}

export type SpotifyClient = ReturnType<typeof initialize>;

export { isError } from './request';
export type { DataPromise, DataResponse } from './types';
export type { AuthProviderOptions };
