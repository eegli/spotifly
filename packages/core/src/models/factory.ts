/* eslint-disable @typescript-eslint/no-explicit-any */
import { chunkify } from '../utils';

/* 
Generate methods for paginated/finite endpoints (e.g. a user's saved
tracks) automatically. Same thing for unpaginated/infinite endpoints
(e.g. multiple tracks, multiple artists)
*/

function fromUnpaginated<
  Func extends (arg1: string[], arg2: any) => Promise<any[]>,
  Options = Parameters<Func>[1] extends infer P
    ? P extends Record<string, any>
      ? P
      : never
    : never,
  ResItem = Awaited<ReturnType<Func>> extends (infer E)[] ? E : never
>({ func, toChunk, limit }: { limit: number; toChunk: string[]; func: Func }) {
  return {
    get: async function get(options?: Options): Promise<ResItem[]> {
      return chunkify(toChunk, limit).reduce(async (acc, idsChunk) => {
        const res = await func(idsChunk, { ...options });
        acc.then(existing => existing.push(...res));
        return acc;
      }, Promise.resolve(<ResItem[]>[]));
    },
    iter: async function* iter(
      options?: {
        chunkSize?: number;
      } & Options
    ): AsyncGenerator<ResItem[]> {
      const chunkSize = options?.chunkSize ?? limit;
      delete options?.chunkSize;

      if (chunkSize < 1) {
        throw new TypeError('Chunk size must be strictly positive');
      }
      if (chunkSize >= toChunk.length) {
        return yield this.get(options);
      }

      let reserve: ResItem[] = [];

      for await (const idsChunk of chunkify(toChunk, limit)) {
        // Append from previous iteration
        const current = reserve;
        reserve = [];
        current.push(...(await func(idsChunk, { ...options })));

        // How many times can we yield given the chunk size and
        // fetched items?
        const yieldableCnt = Math.floor(current.length / chunkSize);

        // We don't have enough items for a chunk yet, continue
        // filling up
        if (yieldableCnt === 0) {
          reserve.push(...current);
          continue;
        }

        // Yield as many times as possible if the chunk size is
        // smaller than the limit. Otherwise, this has no effect
        for (let i = 0; i < yieldableCnt; i++) {
          yield current.slice(i * chunkSize, (i + 1) * chunkSize);
        }

        // Amount of items left that will not fit into a chunk
        const leftoverCnt = current.length % chunkSize;

        // If the leftover count is 0, we were able to yield all
        // currently available items and do not need to transfer
        // leftover items to the next iteration. Otherwise, consider
        // them for the next iteration
        if (leftoverCnt > 0) {
          reserve.push(...current.slice(leftoverCnt * yieldableCnt));
        }
      }

      // The very last chunk of leftovers if there are
      if (reserve.length > 0) return yield reserve;
    },
  };
}

type Paginated = { offset?: number; limit?: number };

function fromPaginated<
  Func extends (arg: any) => Promise<SpotifyApi.PagingObject<any>>,
  Options = Parameters<Func>[number] extends infer P
    ? P extends Paginated
      ? P
      : never
    : never,
  RestOptions = Omit<Options, keyof Paginated>,
  ResAll = Awaited<ReturnType<Func>>,
  Res = ResAll extends SpotifyApi.PagingObject<infer R> ? R : never
>({ func, limit }: { func: Func; limit: number }) {
  async function* iter(options?: RestOptions) {
    let offset = 0;
    let nextPage: string | null = null;
    do {
      const data = await func({ ...options });
      offset += limit;
      nextPage = data.next;
      yield data.items;
    } while (nextPage);
  }

  return {
    get: async function get(
      options?: Options
    ): Promise<SpotifyApi.PagingObject<Res>> {
      const data = await func({ ...options });
      return data;
    },
    getAll: async function getAll(options?: RestOptions): Promise<ResAll[]> {
      const result = [];
      for await (const chunk of iter(options)) {
        result.push(...chunk);
      }
      return result;
    },
  };
}

export default {
  fromPaginated,
  fromUnpaginated,
};
