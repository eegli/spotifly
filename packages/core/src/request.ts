import { AxiosResponse, DataResponse } from './types';

export enum Method {
  GET = 'GET',
  POST = 'POST',
  PUT = 'PUT',
  DELETE = 'DELETE',
}

export function transformResponse<T>(res: AxiosResponse<T>): DataResponse<T> {
  return {
    data: res.data,
    headers: res.headers,
    statusCode: res.status,
  };
}
