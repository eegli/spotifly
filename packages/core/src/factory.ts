import { Params } from './models/params';
import type { AnyObject, AsyncFn, OmitFromAsyncFnParams } from './types';

type PagingObject<T> = SpotifyApi.PagingObject<T>;

type PaginationParams = Pick<Params, 'limit' | 'offset'>;

export function forPaginated<
  F extends AsyncFn<PagingObject<unknown>, string, PaginationParams>,
  R extends Awaited<ReturnType<F>>
>(getFn: F, limit: number) {
  return function (...args: OmitFromAsyncFnParams<F, PaginationParams>) {
    return async function (cb?: (param: R) => unknown): Promise<R[]> {
      const [arg1, arg2] = args;
      let nextPage: string | null = null;
      let offset = 0;
      const responses = [];
      do {
        const response = (await getFn(arg1, {
          ...arg2,
          limit,
          offset,
        })) as R;
        responses.push(response);
        nextPage = response.data.next;
        offset += limit;
        if (cb) await cb(response);
      } while (nextPage);
      return responses;
    };
  };
}

export function forLimited<
  F extends AsyncFn<unknown, string[], AnyObject>,
  R extends Awaited<ReturnType<F>>
>(getFn: F, limit: number) {
  return function (...args: Parameters<F>) {
    const [ids, rest] = args;
    return async function (cb?: (param: R) => unknown): Promise<R[]> {
      const responses: R[] = [];
      for (let i = 0; i < ids.length; i += limit) {
        const chunk = ids.slice(i, i + limit);
        const response = (await getFn(chunk, rest)) as R;
        if (cb) await cb(response);
        responses.push(response);
      }
      return responses;
    };
  };
}
