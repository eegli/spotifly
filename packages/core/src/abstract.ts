import { AuthProvider } from './request';

export abstract class SpotifyKind {
  protected abstract endpoints: Record<string, Record<string, string | number>>;
  protected request: typeof AuthProvider.prototype.request;

  constructor(provider: InstanceType<typeof AuthProvider>) {
    this.request = provider.request.bind(provider);
  }
}
