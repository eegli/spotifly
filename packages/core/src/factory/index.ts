import { AuthProvider } from '../request';

type Endpoint = {
  impl: (provider: AuthProvider) => (...args: any[]) => Promise<any>;
  staticQueryParams?: Record<string, string | number>;
};

export const config: Record<string, Record<string, Endpoint>> = {
  artists: {
    several: {
      staticQueryParams: { limit: 50 },
      impl: provider => async (ids: string[]) => {
        const { data } =
          await provider.request<SpotifyApi.MultipleArtistsResponse>({
            method: 'get',
            url: 'artists',
            params: {
              ids: ids.join(','),
              ...config.artists.several.staticQueryParams,
            },
          });
        return data.artists;
      },
    },
  },
} as const;

const p = new AuthProvider({ accessToken: '' });

config.artists.several
  .impl(p)('1', '2', '3')
  .then(r => r);
