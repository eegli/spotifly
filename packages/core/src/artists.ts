import { RequestAble, SpotifyKind } from './abstract';
import { MapCollection } from './types';
import { chunkify } from './utils';

export class Artists extends SpotifyKind<SpotifyApi.ArtistObjectFull[]> {
  private readonly MULTIPLE_ARTISTS = '/artists';
  private readonly CHUNK_SIZE = 50;

  private collection: MapCollection<SpotifyApi.ArtistObjectFull> = new Map();
  private artistIds: Set<string> = new Set();

  constructor(opts: RequestAble, id?: string, ...ids: string[]) {
    super(opts);
    if (id) {
      this.add(id, ...ids);
    }
  }

  add(id: string, ...ids: string[]) {
    [id, ...ids].forEach(id => this.artistIds.add(id));
  }

  get items() {
    return Array.from(this.collection.values());
  }

  async get(id: string, ...ids: string[]): Promise<void> {
    for await (const chunk of chunkify([id, ...ids], this.CHUNK_SIZE)) {
      const { data } =
        await this.request.get<SpotifyApi.MultipleArtistsResponse>(
          this.BASE_URL_V1 + this.MULTIPLE_ARTISTS,
          {
            params: {
              ids: chunk.join(','),
            },
          }
        );
      data.artists.forEach(artist => {
        this.collection.set(artist.id, artist);
      });
    }
  }
}
