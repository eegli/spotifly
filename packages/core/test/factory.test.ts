import * as factory from '../src/factory';
import { DataPromise } from '../src/request';

beforeEach(() => {
  jest.clearAllMocks();
});

function* createPaginatedBackend(
  total: number,
  atOnce: number
): Generator<
  { nextPage: string | null; items: { id: number }[] },
  void,
  unknown
> {
  const range = Array.from({ length: total }, (_, i) => ({ id: i + 1 }));
  for (let idx = total; idx > 0; idx -= atOnce) {
    yield {
      nextPage: idx > atOnce ? 'next' : null,
      items: range.slice(idx - Math.min(idx, atOnce), idx).reverse(),
    };
  }
}
const setupTest = function (itemCount: number, beLimit: number) {
  const backend = createPaginatedBackend(itemCount, beLimit);

  return jest.fn(
    (
      id: string,
      params?: {
        limit?: number;
        offset?: number;
      }
    ) => {
      let next: string | null = 'next';
      const nextValue = backend.next().value;
      if (!nextValue) throw new Error();
      if (nextValue.nextPage === null) {
        next = null;
      }

      return Promise.resolve({
        statusCode: 200,
        headers: {},
        data: {
          href: 'href',
          items: nextValue.items,
          limit: params?.limit || 20,
          next,
          offset: params?.offset || 0,
          previous: null,
          total: 1,
        },
      });
    }
  );
};

describe('Factory, pagination handling', () => {
  const mockCb = jest.fn((...args: unknown[]) => Promise.resolve(args));
  [
    { limit: 6, itemCount: 20, expectedCalls: 4 },
    { limit: 6, itemCount: 5, expectedCalls: 1 },
    { limit: 6, itemCount: 12, expectedCalls: 2 },
    { limit: 6, itemCount: 1, expectedCalls: 1 },
  ].forEach(testCase => {
    const { limit, itemCount, expectedCalls } = testCase;
    test(`gets all ${itemCount} items and invokes callback in between`, async () => {
      const mockAPI = setupTest(itemCount, limit);

      const fetch = factory.forPaginated(mockAPI.bind(null, 'id'), limit);

      const result = await fetch('id')(args => mockCb(args));

      expect(result).toHaveLength(expectedCalls);
      const allItems = result.reduce(
        (acc, curr) => [...acc, ...curr.data.items],
        [] as unknown[]
      );
      expect(allItems).toHaveLength(itemCount);
      expect(result).toMatchSnapshot('result');

      expect(mockAPI).toHaveBeenCalledTimes(expectedCalls);
      expect(mockAPI.mock.calls).toMatchSnapshot('api invocation');

      expect(mockCb).toHaveBeenCalledTimes(expectedCalls);
      expect(mockCb).toHaveBeenCalledWith(
        expect.objectContaining({
          headers: expect.any(Object),
          statusCode: expect.any(Number),
          data: expect.any(Object),
        })
      );
      expect(mockCb.mock.calls).toMatchSnapshot('callback invocation');
    });
  });
});

const createLimitedAPI = function (limit: number) {
  return jest.fn(
    (
      ids: string[],
      rest?: { market?: string }
    ): DataPromise<{ id: number; market: string }[]> => {
      if (ids.length > limit) {
        throw new Error();
      }

      return Promise.resolve({
        statusCode: 200,
        headers: {},
        data: Array.from({ length: ids.length }).map(() => ({
          id: 1,
          market: rest?.market || 'unknown',
        })),
      });
    }
  );
};

describe('Factory, limited endpoint handling', () => {
  const mockCb = jest.fn((...args: unknown[]) => Promise.resolve(args));

  [
    { limit: 5, itemCount: 1, expectedCalls: 1 },
    { limit: 5, itemCount: 5, expectedCalls: 1 },
    { limit: 5, itemCount: 8, expectedCalls: 2 },
    { limit: 5, itemCount: 12, expectedCalls: 3 },
  ].forEach(testCase => {
    const { limit, itemCount, expectedCalls } = testCase;
    test(`handles max ${itemCount} item with limit of ${limit}`, async () => {
      const mockAPI = createLimitedAPI(limit);
      const fetch = factory.forLimited(mockAPI, limit);

      const result = await fetch(
        Array.from({ length: itemCount }, () => 'id'),
        { market: 'CH' }
      )(args => mockCb(args));
      expect(result).toHaveLength(expectedCalls);
      const allItems = result.reduce(
        (acc, curr) => [...acc, ...curr.data],
        [] as unknown[]
      );
      expect(allItems).toHaveLength(itemCount);
      expect(result).toMatchSnapshot('result');

      expect(mockAPI).toHaveBeenCalledTimes(expectedCalls);
      mockAPI.mock.calls.forEach(call =>
        expect(call[0].length).toBeLessThanOrEqual(limit)
      );

      expect(mockCb).toHaveBeenCalledTimes(expectedCalls);
      expect(mockCb).toHaveBeenCalledWith(
        expect.objectContaining({
          headers: expect.any(Object),
          statusCode: expect.any(Number),
          data: expect.any(Object),
        })
      );
      expect(mockCb.mock.calls).toMatchSnapshot('callback invocation');
    });
  });
});
