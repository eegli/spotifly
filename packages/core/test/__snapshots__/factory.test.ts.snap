// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Factory, limited endpoint handling handles max 1 item with limit of 5: callback invocation 1`] = `
[
  [
    {
      "data": [
        {
          "id": 1,
          "market": "CH",
        },
      ],
      "headers": {},
      "statusCode": 200,
    },
  ],
]
`;

exports[`Factory, limited endpoint handling handles max 1 item with limit of 5: result 1`] = `
[
  {
    "data": [
      {
        "id": 1,
        "market": "CH",
      },
    ],
    "headers": {},
    "statusCode": 200,
  },
]
`;

exports[`Factory, limited endpoint handling handles max 5 item with limit of 5: callback invocation 1`] = `
[
  [
    {
      "data": [
        {
          "id": 1,
          "market": "no market specified",
        },
        {
          "id": 1,
          "market": "no market specified",
        },
        {
          "id": 1,
          "market": "no market specified",
        },
        {
          "id": 1,
          "market": "no market specified",
        },
        {
          "id": 1,
          "market": "no market specified",
        },
      ],
      "headers": {},
      "statusCode": 200,
    },
  ],
]
`;

exports[`Factory, limited endpoint handling handles max 5 item with limit of 5: result 1`] = `
[
  {
    "data": [
      {
        "id": 1,
        "market": "no market specified",
      },
      {
        "id": 1,
        "market": "no market specified",
      },
      {
        "id": 1,
        "market": "no market specified",
      },
      {
        "id": 1,
        "market": "no market specified",
      },
      {
        "id": 1,
        "market": "no market specified",
      },
    ],
    "headers": {},
    "statusCode": 200,
  },
]
`;

exports[`Factory, limited endpoint handling handles max 8 item with limit of 5: callback invocation 1`] = `
[
  [
    {
      "data": [
        {
          "id": 1,
          "market": "CH",
        },
        {
          "id": 1,
          "market": "CH",
        },
        {
          "id": 1,
          "market": "CH",
        },
        {
          "id": 1,
          "market": "CH",
        },
        {
          "id": 1,
          "market": "CH",
        },
      ],
      "headers": {},
      "statusCode": 200,
    },
  ],
  [
    {
      "data": [
        {
          "id": 1,
          "market": "CH",
        },
        {
          "id": 1,
          "market": "CH",
        },
        {
          "id": 1,
          "market": "CH",
        },
      ],
      "headers": {},
      "statusCode": 200,
    },
  ],
]
`;

exports[`Factory, limited endpoint handling handles max 8 item with limit of 5: result 1`] = `
[
  {
    "data": [
      {
        "id": 1,
        "market": "CH",
      },
      {
        "id": 1,
        "market": "CH",
      },
      {
        "id": 1,
        "market": "CH",
      },
      {
        "id": 1,
        "market": "CH",
      },
      {
        "id": 1,
        "market": "CH",
      },
    ],
    "headers": {},
    "statusCode": 200,
  },
  {
    "data": [
      {
        "id": 1,
        "market": "CH",
      },
      {
        "id": 1,
        "market": "CH",
      },
      {
        "id": 1,
        "market": "CH",
      },
    ],
    "headers": {},
    "statusCode": 200,
  },
]
`;

exports[`Factory, limited endpoint handling handles max 12 item with limit of 5: callback invocation 1`] = `
[
  [
    {
      "data": [
        {
          "id": 1,
          "market": "no market specified",
        },
        {
          "id": 1,
          "market": "no market specified",
        },
        {
          "id": 1,
          "market": "no market specified",
        },
        {
          "id": 1,
          "market": "no market specified",
        },
        {
          "id": 1,
          "market": "no market specified",
        },
      ],
      "headers": {},
      "statusCode": 200,
    },
  ],
  [
    {
      "data": [
        {
          "id": 1,
          "market": "no market specified",
        },
        {
          "id": 1,
          "market": "no market specified",
        },
        {
          "id": 1,
          "market": "no market specified",
        },
        {
          "id": 1,
          "market": "no market specified",
        },
        {
          "id": 1,
          "market": "no market specified",
        },
      ],
      "headers": {},
      "statusCode": 200,
    },
  ],
  [
    {
      "data": [
        {
          "id": 1,
          "market": "no market specified",
        },
        {
          "id": 1,
          "market": "no market specified",
        },
      ],
      "headers": {},
      "statusCode": 200,
    },
  ],
]
`;

exports[`Factory, limited endpoint handling handles max 12 item with limit of 5: result 1`] = `
[
  {
    "data": [
      {
        "id": 1,
        "market": "no market specified",
      },
      {
        "id": 1,
        "market": "no market specified",
      },
      {
        "id": 1,
        "market": "no market specified",
      },
      {
        "id": 1,
        "market": "no market specified",
      },
      {
        "id": 1,
        "market": "no market specified",
      },
    ],
    "headers": {},
    "statusCode": 200,
  },
  {
    "data": [
      {
        "id": 1,
        "market": "no market specified",
      },
      {
        "id": 1,
        "market": "no market specified",
      },
      {
        "id": 1,
        "market": "no market specified",
      },
      {
        "id": 1,
        "market": "no market specified",
      },
      {
        "id": 1,
        "market": "no market specified",
      },
    ],
    "headers": {},
    "statusCode": 200,
  },
  {
    "data": [
      {
        "id": 1,
        "market": "no market specified",
      },
      {
        "id": 1,
        "market": "no market specified",
      },
    ],
    "headers": {},
    "statusCode": 200,
  },
]
`;

exports[`Factory, pagination handling gets all 1 items and invokes callback in between: api invocation 1`] = `
[
  [
    "id",
    {
      "limit": 6,
      "market": "CH",
      "offset": 0,
    },
  ],
]
`;

exports[`Factory, pagination handling gets all 1 items and invokes callback in between: callback invocation 1`] = `
[
  [
    {
      "data": {
        "href": "href",
        "items": [
          {
            "id": 1,
          },
        ],
        "limit": 6,
        "next": null,
        "offset": 0,
        "previous": null,
        "total": 1,
      },
      "headers": {},
      "statusCode": 200,
    },
  ],
]
`;

exports[`Factory, pagination handling gets all 1 items and invokes callback in between: result 1`] = `
[
  {
    "data": {
      "href": "href",
      "items": [
        {
          "id": 1,
        },
      ],
      "limit": 6,
      "next": null,
      "offset": 0,
      "previous": null,
      "total": 1,
    },
    "headers": {},
    "statusCode": 200,
  },
]
`;

exports[`Factory, pagination handling gets all 5 items and invokes callback in between: api invocation 1`] = `
[
  [
    "id",
    {
      "limit": 6,
      "market": "CH",
      "offset": 0,
    },
  ],
]
`;

exports[`Factory, pagination handling gets all 5 items and invokes callback in between: callback invocation 1`] = `
[
  [
    {
      "data": {
        "href": "href",
        "items": [
          {
            "id": 5,
          },
          {
            "id": 4,
          },
          {
            "id": 3,
          },
          {
            "id": 2,
          },
          {
            "id": 1,
          },
        ],
        "limit": 6,
        "next": null,
        "offset": 0,
        "previous": null,
        "total": 1,
      },
      "headers": {},
      "statusCode": 200,
    },
  ],
]
`;

exports[`Factory, pagination handling gets all 5 items and invokes callback in between: result 1`] = `
[
  {
    "data": {
      "href": "href",
      "items": [
        {
          "id": 5,
        },
        {
          "id": 4,
        },
        {
          "id": 3,
        },
        {
          "id": 2,
        },
        {
          "id": 1,
        },
      ],
      "limit": 6,
      "next": null,
      "offset": 0,
      "previous": null,
      "total": 1,
    },
    "headers": {},
    "statusCode": 200,
  },
]
`;

exports[`Factory, pagination handling gets all 12 items and invokes callback in between: api invocation 1`] = `
[
  [
    "id",
    {
      "limit": 6,
      "market": "CH",
      "offset": 0,
    },
  ],
  [
    "id",
    {
      "limit": 6,
      "market": "CH",
      "offset": 6,
    },
  ],
]
`;

exports[`Factory, pagination handling gets all 12 items and invokes callback in between: callback invocation 1`] = `
[
  [
    {
      "data": {
        "href": "href",
        "items": [
          {
            "id": 12,
          },
          {
            "id": 11,
          },
          {
            "id": 10,
          },
          {
            "id": 9,
          },
          {
            "id": 8,
          },
          {
            "id": 7,
          },
        ],
        "limit": 6,
        "next": "next",
        "offset": 0,
        "previous": null,
        "total": 1,
      },
      "headers": {},
      "statusCode": 200,
    },
  ],
  [
    {
      "data": {
        "href": "href",
        "items": [
          {
            "id": 6,
          },
          {
            "id": 5,
          },
          {
            "id": 4,
          },
          {
            "id": 3,
          },
          {
            "id": 2,
          },
          {
            "id": 1,
          },
        ],
        "limit": 6,
        "next": null,
        "offset": 6,
        "previous": null,
        "total": 1,
      },
      "headers": {},
      "statusCode": 200,
    },
  ],
]
`;

exports[`Factory, pagination handling gets all 12 items and invokes callback in between: result 1`] = `
[
  {
    "data": {
      "href": "href",
      "items": [
        {
          "id": 12,
        },
        {
          "id": 11,
        },
        {
          "id": 10,
        },
        {
          "id": 9,
        },
        {
          "id": 8,
        },
        {
          "id": 7,
        },
      ],
      "limit": 6,
      "next": "next",
      "offset": 0,
      "previous": null,
      "total": 1,
    },
    "headers": {},
    "statusCode": 200,
  },
  {
    "data": {
      "href": "href",
      "items": [
        {
          "id": 6,
        },
        {
          "id": 5,
        },
        {
          "id": 4,
        },
        {
          "id": 3,
        },
        {
          "id": 2,
        },
        {
          "id": 1,
        },
      ],
      "limit": 6,
      "next": null,
      "offset": 6,
      "previous": null,
      "total": 1,
    },
    "headers": {},
    "statusCode": 200,
  },
]
`;

exports[`Factory, pagination handling gets all 20 items and invokes callback in between: api invocation 1`] = `
[
  [
    "id",
    {
      "limit": 6,
      "market": "CH",
      "offset": 0,
    },
  ],
  [
    "id",
    {
      "limit": 6,
      "market": "CH",
      "offset": 6,
    },
  ],
  [
    "id",
    {
      "limit": 6,
      "market": "CH",
      "offset": 12,
    },
  ],
  [
    "id",
    {
      "limit": 6,
      "market": "CH",
      "offset": 18,
    },
  ],
]
`;

exports[`Factory, pagination handling gets all 20 items and invokes callback in between: callback invocation 1`] = `
[
  [
    {
      "data": {
        "href": "href",
        "items": [
          {
            "id": 20,
          },
          {
            "id": 19,
          },
          {
            "id": 18,
          },
          {
            "id": 17,
          },
          {
            "id": 16,
          },
          {
            "id": 15,
          },
        ],
        "limit": 6,
        "next": "next",
        "offset": 0,
        "previous": null,
        "total": 1,
      },
      "headers": {},
      "statusCode": 200,
    },
  ],
  [
    {
      "data": {
        "href": "href",
        "items": [
          {
            "id": 14,
          },
          {
            "id": 13,
          },
          {
            "id": 12,
          },
          {
            "id": 11,
          },
          {
            "id": 10,
          },
          {
            "id": 9,
          },
        ],
        "limit": 6,
        "next": "next",
        "offset": 6,
        "previous": null,
        "total": 1,
      },
      "headers": {},
      "statusCode": 200,
    },
  ],
  [
    {
      "data": {
        "href": "href",
        "items": [
          {
            "id": 8,
          },
          {
            "id": 7,
          },
          {
            "id": 6,
          },
          {
            "id": 5,
          },
          {
            "id": 4,
          },
          {
            "id": 3,
          },
        ],
        "limit": 6,
        "next": "next",
        "offset": 12,
        "previous": null,
        "total": 1,
      },
      "headers": {},
      "statusCode": 200,
    },
  ],
  [
    {
      "data": {
        "href": "href",
        "items": [
          {
            "id": 2,
          },
          {
            "id": 1,
          },
        ],
        "limit": 6,
        "next": null,
        "offset": 18,
        "previous": null,
        "total": 1,
      },
      "headers": {},
      "statusCode": 200,
    },
  ],
]
`;

exports[`Factory, pagination handling gets all 20 items and invokes callback in between: result 1`] = `
[
  {
    "data": {
      "href": "href",
      "items": [
        {
          "id": 20,
        },
        {
          "id": 19,
        },
        {
          "id": 18,
        },
        {
          "id": 17,
        },
        {
          "id": 16,
        },
        {
          "id": 15,
        },
      ],
      "limit": 6,
      "next": "next",
      "offset": 0,
      "previous": null,
      "total": 1,
    },
    "headers": {},
    "statusCode": 200,
  },
  {
    "data": {
      "href": "href",
      "items": [
        {
          "id": 14,
        },
        {
          "id": 13,
        },
        {
          "id": 12,
        },
        {
          "id": 11,
        },
        {
          "id": 10,
        },
        {
          "id": 9,
        },
      ],
      "limit": 6,
      "next": "next",
      "offset": 6,
      "previous": null,
      "total": 1,
    },
    "headers": {},
    "statusCode": 200,
  },
  {
    "data": {
      "href": "href",
      "items": [
        {
          "id": 8,
        },
        {
          "id": 7,
        },
        {
          "id": 6,
        },
        {
          "id": 5,
        },
        {
          "id": 4,
        },
        {
          "id": 3,
        },
      ],
      "limit": 6,
      "next": "next",
      "offset": 12,
      "previous": null,
      "total": 1,
    },
    "headers": {},
    "statusCode": 200,
  },
  {
    "data": {
      "href": "href",
      "items": [
        {
          "id": 2,
        },
        {
          "id": 1,
        },
      ],
      "limit": 6,
      "next": null,
      "offset": 18,
      "previous": null,
      "total": 1,
    },
    "headers": {},
    "statusCode": 200,
  },
]
`;
